### PR TITLE
Updated reviews.js and index.js in routes and data folders

### PR DESCRIPTION
**Pull request by rgundam-stevens**
Updated reviews.js and index.js in routes and data folders

### CODE REVIEW

**Review by tejasraibagi26 — COMMENTED** [review]
Please use the errorHandlers modules already available for the check. If some error handling is different create new one so others can directly use it from the file instead of redundantly typing it out.

**Review by tejasraibagi26 — COMMENTED** [review]
Please use the errorHandlers modules already available for the check. If some error handling is different create new one so others can directly use it from the file instead of redundantly typing it out.

**Comment by rgundam-stevens** [comment]
I’m not so proficient in coding that’s why I’ve done so 😅

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by rgundam-stevens** [comment]
I’ll change it accordingly by tonight

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: data/reviews.js [hunk 1]
```diff
@@ -0,0 +1,923 @@
...
+  }
+
+  try {
+    ObjectId(venueId);
+  } catch (error) {
+    throw 'Venue Id should be valid ObjectId';
+  }
+
+  let newReview = {
+    _id: ObjectId(),
```
Comment (tejasraibagi26) [inline data/reviews.js:45]: You don't have to create _id, mongodb does it for you automatically.

File: data/reviews.js [hunk 2]
```diff
@@ -0,0 +1,923 @@
...
+    );
+    if (!updateInfo9.matchedCount && !updateInfo9.modifiedCount)
+      throw 'updating dummy _id value for downvoted Reviews failed';
+  }
+
+  const retrivedUser = await userCollection.findOne({ _id: ObjectId(userId) });
+  for (let i = 0; i < retrivedUser.upvotedReviews.length; i++) {
+
+    if (retrivedUser.upvotedReviews[i].id === reviewId) {
+      flag3 = "true";
```
Comment (tejasraibagi26) [inline data/reviews.js:525]: Don't store boolean as string.

File: data/reviews.js [hunk 3]
```diff
@@ -0,0 +1,923 @@
...
+      break;
+    }
+  }
+  if (flag3 != "true") {
```
Comment (tejasraibagi26) [inline data/reviews.js:529]: Kindly change all string value to respective boolean values.

File: data/reviews.js [hunk 4]
```diff
@@ -0,0 +1,923 @@
...
+  } catch (error) {
+    throw 'Review Id should be valid ObjectId';
+  }
+
+  const retrivedReview = await reviewCollection.findOne({ _id: ObjectId(reviewId) });
+  if (!retrivedReview) {
+    throw 'Review not found';
+  }
+
+  let flag1, flag2, flag3, flag4;
```
Comment (tejasraibagi26) [inline data/reviews.js:489]: Atleast give a comment what individual flag does. This is very confusing.

File: routes/reviews.js [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+        return;
+    }
+    try {
+        await venueData.getVenueById(req.params.venueId);
+    } catch (e) {
+        res.status(404).json({ error: 'Venue not found' });
+        return;
+    }
+    try {
+        const updatedReview = await resData.updateReviewText(req.params.id, req.params.userId, req.params.venueId, resInfo.reviewText);
```
Comment (tejasraibagi26) [inline routes/reviews.js:131]: You have extracted all data in their respective variables, so use that instead. Also destructure resInfo as we need to wrap individual params with xss.

File: routes/reviews.js [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+        return;
+    }
+    try {
+        await venueData.getVenueById(req.params.venueId);
+    } catch (e) {
+        res.status(404).json({ error: 'Venue not found' });
+        return;
+    }
+    try {
+        const updatedReview = await resData.updateReviewRating(req.params.id, req.params.userId, req.params.venueId, resInfo.rating);
```
Comment (tejasraibagi26) [inline routes/reviews.js:214]: Follow the same as above.

File: routes/reviews.js [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+        return;
+    }
+    try {
+        await venueData.getVenueById(req.params.venueId);
+    } catch (e) {
+        res.status(404).json({ error: 'Venue not found' });
+        return;
+    }
+    try {
+        await resData.removeReview(req.params.id, req.params.userId, req.params.venueId);
```
Comment (tejasraibagi26) [inline routes/reviews.js:290]: Same as above. Never use req as param input always use the respective variable.

File: routes/reviews.js [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+    }
+    try {
+        await venueData.getVenueById(req.params.venueId);
+    } catch (e) {
+        res.status(404).json({ error: 'Venue not found' });
+        return;
+    }
+
+    try {
+        const postReview = await resData.addReview(req.params.userId, req.params.venueId, resInfo.reviewText, resInfo.rating);
```
Comment (tejasraibagi26) [inline routes/reviews.js:363]: Same as above.

File: routes/reviews.js [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+        return;
+    }
+    try {
+        await userData.getUserById(req.params.userId);
+    } catch (e) {
+        res.status(404).json({ error: 'User not found' });
+        return;
+    }
+    try {
+        const updatedReview = await resData.upVote(req.params.reviewId, req.params.userId);
```
Comment (tejasraibagi26) [inline routes/reviews.js:428]: Same as above.

File: routes/reviews.js [hunk 10]
```diff
@@ -0,0 +1,613 @@
...
+        return;
+    }
+    try {
+        await userData.getUserById(req.params.userId);
+    } catch (e) {
+        res.status(404).json({ error: 'User not found' });
+        return;
+    }
+    try {
+        const updatedReview = await resData.downVote(req.params.reviewId, req.params.userId);
```
Comment (tejasraibagi26) [inline routes/reviews.js:487]: Same as above.

File: routes/reviews.js [hunk 11]
```diff
@@ -0,0 +1,613 @@
...
+        return;
+    }
+    try {
+        await userData.getUserById(req.params.userId);
+    } catch (e) {
+        res.status(404).json({ error: 'User not found' });
+        return;
+    }
+    try {
+        const updatedReview = await resData.removeUpvote(req.params.reviewId, req.params.userId);
```
Comment (tejasraibagi26) [inline routes/reviews.js:547]: Same as above.

File: routes/reviews.js [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+        return;
+    }
+    try {
+        await userData.getUserById(req.params.userId);
+    } catch (e) {
+        res.status(404).json({ error: 'User not found' });
+        return;
+    }
+    try {
+        const updatedReview = await resData.removeDownvote(req.params.reviewId, req.params.userId);
```
Comment (tejasraibagi26) [inline routes/reviews.js:605]: Same as above.

**Review by tejasraibagi26 — CHANGES_REQUESTED** [review]
Please go through the changes.

File: data/reviews.js (hunk 1 above)
Comment (tejasraibagi26) [reply data/reviews.js:45]: Unless it's a sub-doc this is fine.

**Review by tejasraibagi26 — APPROVED** [review]
Everything seems good. I'll wait for Troy to review it and then he can merge it with main.

**Comment by rgundam-stevens** [comment]
Sure ..! Thanks Tejas

File: data/reviews.js [hunk 13]
```diff
@@ -0,0 +1,1078 @@
...
+  /* Updating the review text */
+
+  const updatedInfo = await reviewCollection.updateOne(
+    { _id: ObjectId(id) },
+    { $set: updatedReview }
+  );
+  if (updatedInfo.modifiedCount === 0) {
+    throw 'could not update review text successfully';
+  }
+  const review = await this.getReviewById(id);
```
Comment (troytomasch) [inline data/reviews.js:333]: There is no need for this as it is not used anywhere

File: data/reviews.js [hunk 14]
```diff
@@ -0,0 +1,1078 @@
...
+  if (!venueThatGotPosted) throw 'Venue not found';
+
+  /* Error Handling*/
+
+  let array = [id, userId, venueId, rating];
+  let stringInput = [id, userId, venueId]
+  errorHandler.checkIfElementsExists(array);
+  errorHandler.checkIfElementsAreStrings(stringInput);
+  errorHandler.checkIfElementNotEmptyString(stringInput);
+  errorHandler.checkIfValidRating(rating);
```
Comment (troytomasch) [inline data/reviews.js:380]: Move this up again

File: data/reviews.js [hunk 15]
```diff
@@ -0,0 +1,1078 @@
...
+    {
+      $set: {
+        venueRating: overallRating2
+      }
+    });
+
+  if (!updateInfo2.matchedCount && !updateInfo2.modifiedCount)
+    throw 'updating venue rating failed';
+
+  const review = await this.getReviewById(id);
```
Comment (troytomasch) [inline data/reviews.js:434]: Can remove, this is not used anywhere

File: routes/reviews.js [hunk 16]
```diff
@@ -0,0 +1,615 @@
...
+//---------------------------------------------------------------------------------------------------------
+router.post('/:userId/:venueId', async (req, res) => {
+   
+
+    const userId = req.params.userId;
+    const venueId = req.params.venueId;
+    const reviewText = req.body.reviewText;
+    const rating = req.body.rating;
+
+    let array = [userId, venueId];
```
Comment (troytomasch) [inline routes/reviews.js:308]: This variable is not used

**Review by troytomasch — CHANGES_REQUESTED** [review]
I've made some comments that would be nice if you could fix them. A few of the changes are just making sure you do the error checking before you use the inputs. And also you do a lot of checking of the ObjectId when you could use the function from the error handler. Plus there's some other comments. Just take a look when you get a chance.

**Review by troytomasch — APPROVED** [review]

**Pull request merged**